src/neuroconv/tools/nwb_helpers/_metadata_and_file_helpers.py

```diff
@@ -194,54 +194,73 @@ def make_or_load_nwbfile(
+            backends_that_can_read = [
+                backend_name
+                for backend_name, backend_io_class in BACKEND_NWB_IO.items()
+                if backend_io_class.can_read(path=str(nwbfile_path_in))
+            ]
+            # Future-proofing: raise an error if more than one backend can read the file
+            assert (
+                len(backends_that_can_read) <= 1
+            ), "More than one backend is capable of reading the file! Please raise an issue describing your file."
+            if backend not in backends_that_can_read:
+                raise IOError(
+                    f"The chosen backend ('{backend}') is unable to read the file! "
+                    f"Please select '{backends_that_can_read[0]}' instead."
+                )
         else:
             load_kwargs.update(mode="w")
 
-        backends_that_can_read = [
-            backend_name
-            for backend_name, backend_io_class in BACKEND_NWB_IO.items()
-            if backend_io_class.can_read(path=str(nwbfile_path_in))
-        ]
-        # Future-proofing: raise an error if more than one backend can read the file
-        assert (
-            len(backends_that_can_read) <= 1
-        ), "More than one backend is capable of reading the file! Please raise an issue describing your file."
-        if load_kwargs["mode"] == "r+" and backend not in backends_that_can_read:
-            raise IOError(
-                f"The chosen backend ('{backend}') is unable to read the file! "
-                f"Please select '{backends_that_can_read[0]}' instead."
-            )
-
         io = backend_io_class(**load_kwargs)
 
+    nwbfile_loaded_succesfully = True
+    nwbfile_written_succesfully = True
     try:
         if load_kwargs.get("mode", "") == "r+":
             nwbfile = io.read()
         elif nwbfile is None:
             nwbfile = make_nwbfile_from_metadata(metadata=metadata)
         yield nwbfile
-    except Exception as e:
-        success = False
-        raise e
+    except Exception as load_error:
+        nwbfile_loaded_succesfully = False
+        raise load_error
     finally:
-        if nwbfile_path_in:
+        if nwbfile_path_in is not None and nwbfile_loaded_succesfully:
             try:
-                if success:
-                    io.write(nwbfile)
+                io.write(nwbfile)
 
-                    if verbose:
-                        print(f"NWB file saved at {nwbfile_path_in}!")
+                if verbose:
+                    print(f"NWB file saved at {nwbfile_path_in}!")
+            except Exception as write_error:
+                nwbfile_written_succesfully = False
+                raise write_error
             finally:
                 io.close()
-
-                if not success and not file_initially_exists:
-                    nwbfile_path_in.unlink()
+                del io
+
+        any_load_or_write_error = not nwbfile_loaded_succesfully or not nwbfile_written_succesfully
```

Comment:
question: why did indenting this in the second finally did not work out?

Reply:
Not entirely sure what you're asking

There could be a way of including it in the parent nearby `if` of line 237 but thinking about all the interacting logic gets a bit messy and I figured this was more readable if not more typing

Reply:
The block starting in 250.
I have this impression that first it was within the finally of 246 but that it was moved out.
I am asking if there was a reason for that out of curiosity?

Reply:
I've gone back and forth; it can fit in various levels depending on the logical conditions

I expanded the logic to make a bit more sense in each case, even at the minor cost of repeating the operation multiple times (no harm due to early exit)

Reply:
Ah, OK.  This makes sense.